x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java

```diff
@@ -0,0 +1,119 @@
+import org.elasticsearch.xpack.ql.QlIllegalArgumentException;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Expressions;
+import org.elasticsearch.xpack.ql.expression.Foldables;
+import org.elasticsearch.xpack.ql.expression.Literal;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.sql.type.SqlDataTypeConverter;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isEnum;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isFoldable;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isInteger;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isNumeric;
+
+public abstract class PercentileMethodConfiguration {
+
+    private static final Set<String> ACCEPTED_METHODS =
+        Arrays.stream(PercentilesMethod.values()).map(p -> p.getParseField().getPreferredName()).collect(Collectors.toSet());
+
+    public static Expression defaultMethod(Source source, Expression method) {
+        return method == null
+            ? new Literal(source, PercentilesMethod.TDIGEST.getParseField().getPreferredName(), DataTypes.KEYWORD)
+            : method;
+    }
+
+    public static Expression defaultMethodParameter(Expression methodParameter) {
+        return nullIfUnspecified(methodParameter);
+    }
+
+    public static Expression.TypeResolution resolvePercentileConfiguration(
+        String sourceText, Expression method, Expressions.ParamOrdinal methodOrdinal,
+        Expression methodParameter, Expressions.ParamOrdinal methodParameterOrdinal) {
+
+        Expression.TypeResolution resolution = isEnum(method, sourceText, methodOrdinal, ACCEPTED_METHODS);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (methodParameter != null) {
+            resolution = isFoldable(methodParameter, sourceText, methodParameterOrdinal);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+
+            PercentilesMethod percentilesMethod = asPercentilesMethod(method);
+            switch (percentilesMethod) {
+                case TDIGEST:
+                    resolution = isNumeric(methodParameter, sourceText, methodParameterOrdinal);
+                    break;
+                case HDR:
+                    resolution = isInteger(methodParameter, sourceText, methodParameterOrdinal);
+                    break;
+                default:
+                    throw new QlIllegalArgumentException("Not handled PercentilesMethod [" + percentilesMethod + "]");
+            }
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
```

Comment:
Why is there a need for `isEnum` if the methodParam is checked again to be folded and then checked against the enum values itself?
Further more the point of type resolution is to *not* throw exceptions but return error messages.